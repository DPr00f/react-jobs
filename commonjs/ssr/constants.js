'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/* eslint-disable import/prefer-default-export */

var STATE_IDENTIFIER = exports.STATE_IDENTIFIER = '__REACT_JOBS_STATE__';