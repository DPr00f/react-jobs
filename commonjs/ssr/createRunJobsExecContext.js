"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createRenderContext;
function createRenderContext() {
  var jobsState = {};
  return {
    registerJobState: function registerJobState(jobID, state) {
      jobsState[jobID] = state;
    },
    getState: function getState() {
      return {
        jobsState: jobsState
      };
    }
  };
}