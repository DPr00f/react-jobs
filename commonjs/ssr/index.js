'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rehydrateJobs = exports.runJobs = exports.withJob = undefined;

var _withJob = require('./withJob');

var _withJob2 = _interopRequireDefault(_withJob);

var _runJobs = require('./runJobs');

var _runJobs2 = _interopRequireDefault(_runJobs);

var _rehydrateJobs = require('./rehydrateJobs');

var _rehydrateJobs2 = _interopRequireDefault(_rehydrateJobs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.withJob = _withJob2.default;
exports.runJobs = _runJobs2.default;
exports.rehydrateJobs = _rehydrateJobs2.default;