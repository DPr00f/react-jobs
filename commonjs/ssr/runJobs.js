'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = runJobs;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactTreeWalker = require('react-tree-walker');

var _reactTreeWalker2 = _interopRequireDefault(_reactTreeWalker);

var _utils = require('../utils');

var _ServerProvider = require('./ServerProvider');

var _ServerProvider2 = _interopRequireDefault(_ServerProvider);

var _createRunJobsExecContext = require('./createRunJobsExecContext');

var _createRunJobsExecContext2 = _interopRequireDefault(_createRunJobsExecContext);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getJobs(rootElement, rootContext) {
  var fetchRoot = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var jobs = [];

  (0, _reactTreeWalker2.default)(rootElement, function (element, instance, context) {
    var skipRoot = !fetchRoot && element === rootElement;
    if (instance && typeof instance.getExecutingJob === 'function' && !skipRoot) {
      var job = instance.getExecutingJob();
      if ((0, _utils.isPromise)(job)) {
        jobs.push({ job: job, element: element, context: context });

        // Tell walkTree to not recurse inside this component;  we will
        // wait for the query to execute before attempting it.
        return false;
      }
    }

    return undefined;
  }, rootContext);

  return jobs;
}

function runJobs(rootElement) {
  var rootContext = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var isRoot = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var processingElement = void 0;
  if (isRoot) {
    var runJobsExecContext = (0, _createRunJobsExecContext2.default)();
    rootContext.runJobsExecContext = runJobsExecContext; // eslint-disable-line no-param-reassign
    processingElement = _react2.default.createElement(
      _ServerProvider2.default,
      { runJobsExecContext: runJobsExecContext },
      rootElement
    );
  } else {
    processingElement = rootElement;
  }

  var resolveResult = function resolveResult() {
    return {
      appWithJobs: processingElement,
      state: rootContext.runJobsExecContext.getState(),
      STATE_IDENTIFIER: _constants.STATE_IDENTIFIER
    };
  };

  var runningJobs = getJobs(processingElement, rootContext, isRoot)
  // Map over the jobs and then...
  .map(function (_ref) {
    var job = _ref.job,
        element = _ref.element,
        context = _ref.context;
    return (
      // ... make sure they will continue executing jobs for their Children
      // when they are complete.
      job.then(function () {
        return runJobs(element, context,
        // We've just grabbed the job for element so don't try and get it again
        false);
      })
    );
  });

  return runningJobs.length > 0 ? Promise.all(runningJobs)
  // Swallow errors.
  .catch(function () {
    return undefined;
  }).then(function () {
    return resolveResult();
  }) : Promise.resolve(resolveResult());
}