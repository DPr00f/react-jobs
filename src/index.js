/* @flow */

import JobsProvider from './JobsProvider';
import job from './job';

export default {
  JobsProvider,
  job,
};
