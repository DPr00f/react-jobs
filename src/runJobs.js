/* @flow */

import { Children, Element } from 'react';
import { isPromise } from './utils';

type React$Element = Element<*>;
type Context = { [key: string]: any; };
type ElementVisitor =
  (element: React$Element, instance: ?Function, context: Context) => boolean | void;
type ElementJob = {
  job: Promise<any>,
  element: React$Element,
  context: Context,
}

// Recurse an React Element tree, running visitor on each element.
// If visitor returns `false`, don't call the element's render function
//   or recurse into its child elements
export function walkTree(
  element: React$Element,
  context: Context,
  visitor: ElementVisitor,
) {
  const Component = element.type;

  // a stateless functional component or a class
  if (typeof Component === 'function') {
    const props = Object.assign({}, Component.defaultProps, element.props);
    let childContext = context;
    let child;

    // Are we are a react class?
    // http://bit.ly/2j9Ifk3
    const isReactClassComponent = Component.prototype &&
      (Component.prototype.isReactComponent || Component.prototype.isPureReactComponent);

    if (isReactClassComponent) {
      const instance = new Component(props, context);
      // In case the user doesn't pass these to super in the constructor
      instance.props = instance.props || props;
      instance.context = instance.context || context;

      // Override setState to just change the state, not queue up an update.
      //   (we can't do the default React thing as we aren't mounted "properly"
      //   however, we don't need to re-render as well only support setState in
      //   componentWillMount, which happens *before* render).
      instance.setState = (newState) => {
        instance.state = Object.assign({}, instance.state, newState);
      };

      if (instance.componentWillMount) {
        instance.componentWillMount();
      }

      if (instance.getChildContext) {
        childContext = Object.assign({}, context, instance.getChildContext());
      }

      if (visitor(element, instance, context) === false) {
        return;
      }

      child = instance.render();
    } else { // just a stateless functional
      if (visitor(element, null, context) === false) {
        return;
      }

      child = Component(props, context);
    }

    if (child) {
      walkTree(child, childContext, visitor);
    }
  } else { // a basic string or dom element, just get children
    if (visitor(element, null, context) === false) {
      return;
    }

    if (element.props && element.props.children) {
      Children.forEach(element.props.children, (child: any) => {
        if (child) {
          walkTree(child, context, visitor);
        }
      });
    }
  }
}

function getJobs(
  rootElement : React$Element,
  rootContext : Context,
  fetchRoot : boolean = true,
) : ElementJob[] {
  const jobs = [];

  walkTree(rootElement, rootContext, (element, instance, context) => {
    const skipRoot = !fetchRoot && (element === rootElement);
    if (instance && typeof instance.getExecutingJob === 'function' && !skipRoot) {
      const job = instance.getExecutingJob();
      if (isPromise(job)) {
        jobs.push({ job, element, context });

        // Tell walkTree to not recurse inside this component;  we will
        // wait for the query to execute before attempting it.
        return false;
      }
    }

    return undefined;
  });

  return jobs;
}

export default function runJobs(
  rootElement : React$Element,
  rootContext : Context = {},
  fetchRoot : boolean = true,
) {
  const jobs = getJobs(rootElement, rootContext, fetchRoot);

  // no queries found, nothing to do
  if (!jobs.length) {
    return Promise.resolve(null);
  }

  // wait on each query that we found, re-rendering the subtree when it's done
  const mappedJobs = jobs.map(({ job, element, context }) =>
    // we've just grabbed the query for element so don't try and get it again
    job.then(() => runJobs(element, context, false)),
  );

  return Promise.all(mappedJobs).then(() => null);
}
