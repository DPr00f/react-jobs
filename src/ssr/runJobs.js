/* @flow */

// eslint-disable-next-line import/no-extraneous-dependencies
import React, { Element } from 'react'
// eslint-disable-next-line import/no-extraneous-dependencies
import reactTreeWalker from 'react-tree-walker'
import { isPromise } from '../utils'
import ServerProvider from './ServerProvider'
import createRunJobsExecContext from './createRunJobsExecContext'
import { STATE_IDENTIFIER } from './constants'
import type { RehydrateState } from './types'

type React$Element = Element<*>;
type Context = { [key: string]: any };
type ElementJob = {
  job: Promise<any>,
  element: React$Element,
  context: Context,
};
type RunJobsResult = {
  appWithJobs: React$Element,
  state: RehydrateState,
  STATE_IDENTIFIER: string,
};

function getJobs(
  rootElement: React$Element,
  rootContext: Object,
  fetchRoot: boolean = true,
): ElementJob[] {
  const jobs = []

  reactTreeWalker(
    rootElement,
    (element, instance, context) => {
      const skipRoot = !fetchRoot && element === rootElement
      if (
        instance && typeof instance.getExecutingJob === 'function' && !skipRoot
      ) {
        const job = instance.getExecutingJob()
        if (isPromise(job)) {
          jobs.push({ job, element: instance, context })

          // Tell walkTree to not recurse inside this component;  we will
          // wait for the query to execute before attempting it.
          return false
        }
      }

      return undefined
    },
    rootContext,
  )

  return jobs
}

export default function runJobs(
  rootElement: Element<any>,
  rootContext: Object = {},
  isRoot: boolean = true,
): Promise<RunJobsResult> {
  let processingElement
  if (isRoot) {
    const runJobsExecContext = createRunJobsExecContext()
    rootContext.runJobsExecContext = runJobsExecContext // eslint-disable-line no-param-reassign
    processingElement = (
      <ServerProvider runJobsExecContext={runJobsExecContext}>
        {rootElement}
      </ServerProvider>
    )
  } else {
    processingElement = rootElement
  }

  const resolveResult = () => ({
    appWithJobs: processingElement,
    state: rootContext.runJobsExecContext.getState(),
    STATE_IDENTIFIER,
  })

  const runningJobs = getJobs(processingElement, rootContext, isRoot)
    // Map over the jobs and then...
    .map(({ job, element, context }) =>
      // ... make sure they will continue executing jobs for their Children
      // when they are complete.
      job.then(() =>
        runJobs(
          element,
          context,
          // We've just grabbed the job for element so don't try and get it again
          false,
        )))

  return runningJobs.length > 0
    ? Promise.all(runningJobs)
        // Swallow errors.
        .catch(() => undefined)
        .then(() => resolveResult())
    : Promise.resolve(resolveResult())
}
