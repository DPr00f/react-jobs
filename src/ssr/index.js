/*  */

import withJob from './withJob'
import runJobs from './runJobs'
import rehydrateJobs from './rehydrateJobs'

export { withJob, runJobs, rehydrateJobs }
