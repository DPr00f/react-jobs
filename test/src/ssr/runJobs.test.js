/* @flow */

import React from 'react';
import { mount } from 'enzyme';
import { job } from '../../../src/ssr';
import { Foo, resolveAfter } from '../../helpers';

// Under test.
import runJobs from '../../../src/ssr/runJobs';

const workTime = 10; // ms

describe('runJobs()', () => {
  it('should return a Promise', () => {
    const actual = runJobs(<div>foo</div>);
    expect(actual.then).toBeTruthy();
  });

  it('should render simple job with success', () => {
    const FooWithJob = job(() => resolveAfter(workTime, 'Hello world!'))(Foo);
    const app = <FooWithJob />;
    return runJobs(app).then(({ app: wrappedApp }) => {
      expect(mount(wrappedApp)).toMatchSnapshot();
    });
  });
});
