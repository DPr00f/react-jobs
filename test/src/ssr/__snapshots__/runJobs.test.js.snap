exports[`runJobs() should not render a job that is deferred 1`] = `
<ServerProvider
  runJobsExecContext={
    Object {
      "getState": [Function],
      "registerJobState": [Function],
    }
  }>
  <ClientProvider>
    <FooWithJobWithJobID>
      <FooWithJob
        jobInitState={
          Object {
            "completed": false,
            "inProgress": true,
          }
        }
        onJobProcessed={[Function]}>
        <Foo
          job={
            Object {
              "completed": false,
              "error": undefined,
              "inProgress": true,
              "result": undefined,
            }
          }
          jobInitState={
            Object {
              "completed": false,
              "inProgress": true,
            }
          }
          onJobProcessed={[Function]}>
          <div>
            In progress...
          </div>
        </Foo>
      </FooWithJob>
    </FooWithJobWithJobID>
  </ClientProvider>
</ServerProvider>
`;

exports[`runJobs() should render a job that fails 1`] = `
<ServerProvider
  runJobsExecContext={
    Object {
      "getState": [Function],
      "registerJobState": [Function],
    }
  }>
  <ClientProvider>
    <FooWithJobWithJobID>
      <FooWithJob
        jobInitState={
          Object {
            "completed": true,
            "error": "Poop!",
            "inProgress": false,
            "result": undefined,
          }
        }
        onJobProcessed={[Function]}>
        <Foo
          job={
            Object {
              "completed": true,
              "error": "Poop!",
              "inProgress": false,
              "result": undefined,
            }
          }
          jobInitState={
            Object {
              "completed": true,
              "error": "Poop!",
              "inProgress": false,
              "result": undefined,
            }
          }
          onJobProcessed={[Function]}>
          <div>
            Error: 
            Poop!
          </div>
        </Foo>
      </FooWithJob>
    </FooWithJobWithJobID>
  </ClientProvider>
</ServerProvider>
`;

exports[`runJobs() should render a job that succeeds 1`] = `
<ServerProvider
  runJobsExecContext={
    Object {
      "getState": [Function],
      "registerJobState": [Function],
    }
  }>
  <ClientProvider>
    <FooWithJobWithJobID>
      <FooWithJob
        jobInitState={
          Object {
            "completed": true,
            "error": undefined,
            "inProgress": false,
            "result": "Hello world!",
          }
        }
        onJobProcessed={[Function]}>
        <Foo
          job={
            Object {
              "completed": true,
              "error": undefined,
              "inProgress": false,
              "result": "Hello world!",
            }
          }
          jobInitState={
            Object {
              "completed": true,
              "error": undefined,
              "inProgress": false,
              "result": "Hello world!",
            }
          }
          onJobProcessed={[Function]}>
          <div>
            Hello world!
          </div>
        </Foo>
      </FooWithJob>
    </FooWithJobWithJobID>
  </ClientProvider>
</ServerProvider>
`;
